lib/Runtime/Language/JavascriptOperators.cpp

```diff
@@ -1091,10 +1090,32 @@ namespace Js
             }
             else
             {
-                return object && object->HasOwnProperty(propertyId);
+#ifdef DEBUG
+                // In debug builds, calculate expectedAnswer up-front so that
+                // we can Assert the fastpath gets the right result.
+                BOOL expectedAnswer = object && object->HasOwnProperty(propertyId);
+#endif
+
+                PropertyString *propString = requestContext->TryGetPropertyString(propertyId);
+                if (propString != nullptr)
+                {
+                    const PropertyCache *propCache = propString->GetPropertyCache();
+                    if (object->GetType() == propCache->type)
+                    {
+                        // The type cached for the property was the same as the type of this object
+                        // (i.e. obj in obj.hasOwnProperty), so we know the answer is "true".
+                        Assert(expectedAnswer == TRUE); // sanity check on the fastpath result
```

Comment:
Could be nicer to remove all `ifdef DEBUG` lines above and put `Assert(object && object->HasOwnProperty(propertyId))` here instead of `expectedAnswer == TRUE`